// @uploadcare/upload-client/lib/api/fromUrl.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isFileInfoResponse = exports.isTokenResponse = exports.TypeEnum = void 0;
var request_node_1 = require("../request/request.node");
var getUrl_1 = require("../tools/getUrl");
var defaultSettings_1 = require("../defaultSettings");
var userAgent_1 = require("../tools/userAgent");
var camelizeKeys_1 = require("../tools/camelizeKeys");
var errors_1 = require("../tools/errors");
var retryIfThrottled_1 = require("../tools/retryIfThrottled");
var TypeEnum;
(function (TypeEnum) {
    TypeEnum["Token"] = "token";
    TypeEnum["FileInfo"] = "file_info";
})(TypeEnum = exports.TypeEnum || (exports.TypeEnum = {}));
/**
 * TokenResponse Type Guard.
 */
exports.isTokenResponse = function (response) {
    return response.type !== undefined && response.type === TypeEnum.Token;
};
/**
 * FileInfoResponse Type Guard.
 */
exports.isFileInfoResponse = function (response) {
    return response.type !== undefined && response.type === TypeEnum.FileInfo;
};
/**
 * Uploading files from URL.
 */
function fromUrl(sourceUrl, _a) {
    var publicKey = _a.publicKey, _b = _a.baseURL, baseURL = _b === void 0 ? defaultSettings_1.default.baseURL : _b, store = _a.store, fileName = _a.fileName, checkForUrlDuplicates = _a.checkForUrlDuplicates, saveUrlForRecurrentUploads = _a.saveUrlForRecurrentUploads, secureSignature = _a.secureSignature, secureExpire = _a.secureExpire, _c = _a.source, source = _c === void 0 ? 'url' : _c, cancel = _a.cancel, integration = _a.integration, _d = _a.retryThrottledRequestMaxTimes, retryThrottledRequestMaxTimes = _d === void 0 ? defaultSettings_1.default.retryThrottledRequestMaxTimes : _d;
    return retryIfThrottled_1.default(function () {
        return request_node_1.default({
            method: 'POST',
            headers: {
                'X-UC-User-Agent': userAgent_1.getUserAgent({ publicKey: publicKey, integration: integration })
            },
            url: getUrl_1.default(baseURL, '/from_url/', {
                jsonerrors: 1,
                pub_key: publicKey,
                source_url: sourceUrl,
                store: typeof store === 'undefined' ? 'auto' : store ? 1 : undefined,
                filename: fileName,
                check_URL_duplicates: checkForUrlDuplicates ? 1 : undefined,
                save_URL_duplicates: saveUrlForRecurrentUploads ? 1 : undefined,
                signature: secureSignature,
                expire: secureExpire,
                source: source
            }),
            cancel: cancel
        }).then(function (_a) {
            var data = _a.data, headers = _a.headers, request = _a.request;
            var response = camelizeKeys_1.default(JSON.parse(data));
            if ('error' in response) {
                throw new errors_1.UploadClientError("[" + response.error.statusCode + "] " + response.error.content, request, response.error, headers);
            }
            else {
                return response;
            }
        });
    }, retryThrottledRequestMaxTimes);
}
exports.default = fromUrl;
